src/python/pants/bsp/goal.py

```diff
@@ -26,15 +38,140 @@ class BSPGoal(BuiltinGoal):
     def activated(cls, union_membership: UnionMembership) -> bool:
         return False
 
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--server",
+            type=bool,
+            default=False,
+            advanced=True,
+            help=(
+                "Run the Build Server Protocol server. Pants will receive RPC requests via the console. "
+                "This should only be invoked via the IDE."
+            ),
+        )
+        register(
+            "--runner-env-vars",
+            type=list,
+            member_type=str,
+            default=["PATH"],
+            help=(
+                "Environment variables to set in the BSP runner script when setting up BSP support. "
+                "Entries are either strings in the form `ENV_VAR=value` to set an explicit value; "
+                "or just `ENV_VAR` to copy the value from Pants' own environment.\n\n"
+                "Note: IntelliJ on macOS apparently does not pass through the PATH from the shell, "
+                "and so on macOS at the very least, `PATH` should be explicitly written into the BSP "
+                "runner script via this option."
+            ),
+            advanced=True,
+        )
+
     def run(
         self,
         *,
         build_config: BuildConfiguration,
         graph_session: GraphSession,
         options: Options,
         specs: Specs,
-        union_membership: UnionMembership
+        union_membership: UnionMembership,
+    ) -> ExitCode:
+        goal_options = options.for_scope(self.name)
+        if goal_options.server:
+            return self._run_server(
+                graph_session=graph_session,
+                union_membership=union_membership,
+            )
+        current_session_values = graph_session.scheduler_session.py_session.session_values
+        env = current_session_values[CompleteEnvironment]
+        return self._setup_bsp_connection(
+            union_membership=union_membership, env=env, options=goal_options
+        )
+
+    def _setup_bsp_connection(
+        self,
+        union_membership: UnionMembership,
+        env: Mapping[str, str],
+        options: OptionValueContainer,
     ) -> ExitCode:
+        """Setup the BSP connection file."""
+
+        build_root = BuildRoot()
+        bsp_conn_path = build_root.pathlib_path / ".bsp" / "pants.json"
+        if bsp_conn_path.exists():
+            print(
+                f"ERROR: A BSP connection file already exists at path `{bsp_conn_path}`. "
+                "Please delete that file if you intend to re-setup BSP in this repository."
```

Comment:
Or should it be `logger.error()`? I'm not sure how builtin goals work.

Reply:
It's more of a UX question. Should this code be raising an exception, logging to stderr, or printing to stderr?

Note: I had tried using `console.write_stderr`, but that failed with an exception related to SideEffecting not being allowed in non goal rules. (But this is outside the engine, so using `Console` would seem to be okay.) The help code just `print`s to I went that way but those messages are not errors, so probably not the right pattern to copy.)

Reply:
returning `PANTS_FAILED_EXIT_CODE` is the right approach. Printing the error seems fine to me because things like timestamp are irrelevant, just like we wouldn't want timestamp for `help`.